### PR TITLE
SilverStripe 4.0 Compatability

### PR DESCRIPTION
**Pull request by brettt89**
Update module to be compatible with SilverStripe 4.0

### CODE REVIEW

File: src/Services/QueuedJobService.php [hunk 1]
```diff
@@ -680,7 +681,7 @@ public function runJob($jobId)
                 }
             } catch (Exception $e) {
                 // okay, we'll just catch this exception for now
-                $this->getLogger()->error($e->getMessage());
+                $this->getLogger()->debug($e->getMessage());
```
Comment (sminnee) [inline src/Services/QueuedJobService.php]: Why have you changed the level of these messages?

Comment (brettt89) [inline src/Services/QueuedJobService.php]: I had done so after reading https://docs.silverstripe.org/en/4/developer_guides/debugging/error_handling/

However I misread it as "debug and warn" should be through logger and "user_error" for errors.

**Comment by brettt89** [comment]
Have been running through some tests manually and have found a couple of bugs that don't look to be covered by tests. Working through those now.

**Comment by brettt89** [comment]
Date fields weren't updated. Updating these and adding tests for checking correct dates.

**Comment by edlinklater** [comment]
Fixes #126

**Comment by brettt89** [comment]
Fixes #122

File: tests/QueuedJobsTest/TestQJService.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace SilverStripe\QueuedJobs\Tests;
+
+use SilverStripe\QueuedJobs\Services\QueuedJobService;
+
+// stub class to be able to call init from an external context
+class TestQJService extends QueuedJobService
```
Comment (robbieaverill) [inline tests/QueuedJobsTest/TestQJService.php]: These stubs should probably implement `SilverStripe\Dev\TestOnly` (I know they didn't before)

File: tests/ScheduledExecutionTest.php [hunk 3]
```diff
@@ -118,9 +125,10 @@ public function testScheduledExecutionInterval()
     }
 }
 
-
-class TestScheduledDataObject extends DataObject implements TestOnly
+class TestScheduledDataObject extends DataObject
```
Comment (robbieaverill) [inline tests/ScheduledExecutionTest.php]: Add test only back

Comment (brettt89) [inline tests/ScheduledExecutionTest.php]: If I add "TestOnly" to these objects, the tests fail as SapphireTest ignores buildings the database for these objects.

https://github.com/silverstripe/silverstripe-framework/blob/master/src/Dev/SapphireTest.php#L1173

And the DB table is required due to the write occuring in the test
https://github.com/brettt89/silverstripe-queuedjobs/blob/compat-4.0/tests/ScheduledExecutionTest.php#L33

File: tests/QueuedJobsTest/TestQJService.php (hunk 2 above)
Comment (brettt89) [inline tests/QueuedJobsTest/TestQJService.php]: If I add "TestOnly" to these objects, the tests fail as SapphireTest ignores buildings the database for these objects.

https://github.com/silverstripe/silverstripe-framework/blob/master/src/Dev/SapphireTest.php#L1173

File: tests/ScheduledExecutionTest.php (hunk 3 above)
Comment (robbieaverill) [inline tests/ScheduledExecutionTest.php]: Sounds like a bug to me!

File: tests/ScheduledExecutionTest.php [hunk 4]
```diff
@@ -11,11 +11,18 @@
...
+     *
+     * @var bool
+     */
+    protected $usesDatabase = true;
+
+    /**
      * {@inheritDoc}
      * @var array
      */
     protected $extraDataObjects = array(
```
Comment (sminnee) [inline tests/ScheduledExecutionTest.php]: this should be `protected static $extra_dataobjects`

**Comment by brettt89** [comment]
@sminnee @robbieaverill Have updated tests to correctly implement TestOnly.

Warnings in tests are due to expected errors thrown by QueuedJobs.

**Comment by sminnee** [comment]
OK those warnings block merge. We'll need to think of a better way of handling these.

Perhaps the tests need to mock in a logger than can pick up these issues and validate that they have appeared?

**Comment by robbieaverill** [comment]
Yeah, you can push a NullHandler from monolog to the logger from the setUp method so it won't default to logging these to the terminal

**Comment by brettt89** [comment]
@sminnee @robbieaverill Have pushed a NullHandler onto the ErrorHandler using Injector. PHP is not reporting them as "Warnings", however the messages are still appearing in the output. Is that going to be an issue?

**Comment by sminnee** [comment]
It makes it look broken, which inhibits all future dev, so yes I think we want to sort it out.

**Comment by sminnee** [comment]
Are you sure that the messages are, in fact, benign?

File: src/QJUtils.php [hunk 5]
```diff
@@ -5,6 +5,7 @@
 use SilverStripe\Core\Convert;
 use SilverStripe\Core\Injector\Injector;
 use SilverStripe\ORM\DB;
+use Monolog\Logger;
```
Comment (robbieaverill) [inline src/QJUtils.php]: Is this used in the class?

File: .travis.yml [hunk 6]
```diff
@@ -5,7 +5,6 @@ sudo: false
 language: php
 
 php:
-  - 5.5
```
Comment (robbieaverill) [inline .travis.yml]: May as well add 7.1 while you're there

Comment (brettt89) [inline .travis.yml]: Added in commit 2af7805bbc37d459e0a829cda2867cf0aed7eeac

**Comment by brettt89** [comment]
@sminnee Have updated QueuedJobsTest to catch errors and check that expected errors exist after specific tests. Have confirmed that errors are expected as it is testing how the system reacts after failure.

@robbieaverill Have removed reference to Monologger/Logger and added tests for 7.1

**Comment by nyeholt** [comment]
@sminnee did Brett's last update resolve the issue you had?

**Comment by sminnee** [comment]
Yeah I'm happy with this now — no gunk in the test output.

I'd note that we probably don't need so many build configs and that the following should be sufficient:

 * PHP 5.6 + MYSQL
 * PHP 7.0 + PGSQL
 * PHP 7.1 + MYSQL

There's a finite number of concurrent builds across all `github.com/silverstripe-australia` runs so it's good to keep your job counts in check. ;-)

**Comment by robbieaverill** [comment]
Assuming this PR fixes #122 (it's noted above as doing so, and the builds pass so I assume that's true) then it replaces #123 too (also replaced by #124). I've closed #123.

**Pull request merged**